src/query/service/src/interpreters/interpreter_execute_immediate.rs

```diff
@@ -0,0 +1,425 @@
+use databend_common_script::Executor;
+use databend_common_script::ReturnValue;
+use databend_common_sql::plans::ExecuteImmediatePlan;
+use databend_common_sql::Planner;
+use databend_common_storages_fuse::TableContext;
+use futures::TryStreamExt;
+use itertools::Itertools;
+use serde_json::Value as JsonValue;
+
+use crate::interpreters::Interpreter;
+use crate::interpreters::InterpreterFactory;
+use crate::pipelines::PipelineBuildResult;
+use crate::sessions::QueryContext;
+
+#[derive(Debug)]
+pub struct ExecuteImmediateInterpreter {
+    ctx: Arc<QueryContext>,
+    plan: ExecuteImmediatePlan,
+}
+
+impl ExecuteImmediateInterpreter {
+    pub fn try_create(ctx: Arc<QueryContext>, plan: ExecuteImmediatePlan) -> Result<Self> {
+        Ok(ExecuteImmediateInterpreter { ctx, plan })
+    }
+}
+
+#[async_trait::async_trait]
+impl Interpreter for ExecuteImmediateInterpreter {
+    fn name(&self) -> &str {
+        "ExecuteImmediateInterpreter"
+    }
+
+    fn is_ddl(&self) -> bool {
+        false
+    }
+
+    #[minitrace::trace]
+    #[async_backtrace::framed]
+    async fn execute2(&self) -> Result<PipelineBuildResult> {
+        let res: Result<_> = try {
+            let dialect = self.ctx.get_settings().get_sql_dialect()?;
+            let tokens = tokenize_sql(&self.plan.script)?;
+            let mut ast = run_parser(&tokens, dialect, ParseMode::Template, false, script_block)?;
+
+            let mut src = vec![];
+            for declare in ast.declares {
+                match declare {
+                    DeclareItem::Var(declare) => src.push(ScriptStatement::LetVar { declare }),
+                    DeclareItem::Set(declare) => {
+                        src.push(ScriptStatement::LetStatement { declare })
+                    }
+                }
+            }
+            src.append(&mut ast.body);
+            let compiled = compile(&src)?;
+
+            let client = ScriptClient {
+                ctx: self.ctx.clone(),
+            };
+            let mut executor = Executor::load(ast.span, client, compiled);
+            let result = executor.run(/* TODO(andylokandy) */ 1000).await?;
+
+            match result {
+                Some(ReturnValue::Var(scalar)) => {
+                    PipelineBuildResult::from_blocks(vec![DataBlock::new_from_columns(vec![
+                        StringType::from_data(vec![scalar.to_string()]),
+                    ])])?
+                }
+                Some(ReturnValue::Set(set)) => {
+                    let rendered_table = box_render(
+                        &set.schema,
+                        &[set.block.clone()],
+                        usize::MAX,
+                        usize::MAX,
+                        usize::MAX,
+                        true,
+                    )?;
+                    let lines = rendered_table.lines().map(|x| x.to_string()).collect();
+                    PipelineBuildResult::from_blocks(vec![DataBlock::new_from_columns(vec![
```

Comment:
The output result is a single-string column now.

Can we infer the schema before execution?

Reply:
No, there is no way to infer the schema. Unlike EXECUTE IMMEDIATE, CREATE PROCEDURE must explicitly declare its returning schema.